client/signup/index.node.js

```diff
@@ -3,38 +3,19 @@
 /**
  * Internal dependencies
  */
+import { getLanguage, getLanguageSlugs } from 'lib/i18n-utils';
 
-import { getLanguage } from 'lib/i18n-utils';
+const lang = `:lang(${ getLanguageSlugs().join( '|' ) })?`;
 
 export default function( router ) {
-	router( '/start/:flowName?/:stepName?/:stepSectionName?/:lang?', setUpLocale );
+	router( `/start/:flowName?/:stepName?/:stepSectionName?/${ lang }`, setUpLocale );
```

Comment:
I tell you what I find confusing: multiple optional parameters in a row. It makes me wonder what happens when only the language slug is there. Would it ever thing it's a slug or would it always think it's a flow name?

Can we additionally clean this up by removing the ambiguity with a few extra routes…

```js
router( `/start/:flowName/:stepName/:stepSectionName/${ lang }`, setUpLocale );
router( `/start/:flowName?/:stepName?/:stepSectionName`, setUpLocale );
router( `/start/:flowName?/:stepName`, setUpLocale );
router( `/start/:flowName?, setUpLocale );
```

Reply:
Agreed, this is nearly impossible to reason about.

Reply:
Great idea! I _think_ those routes will need to looks slightly different (all of them need to contain `${ lang }`), but I think we'll get there...

Reply:
55ced49

Reply:
> (all of them need to contain ${ lang })

this is the unavoidable ambiguity then - what happens is that the _parse_ becomes unambiguous but the _grammar_ is.

```js
const paths = [
	'/start/flow-a/step-b/de_DE',
	'/start/flow-a/step-b/section-c',
];
```

In the above two paths what is last parameter supposed to represent? If we encode in the parser that it must be a section name if there are only three args then we can't really make it understand the locale code. If we try to fit the optional locale code in then we can't make an accurate parse decision.

If we want to avoid this we're going to have to make the parser attempt to parse the locale codes specifically and only try the next item if the pattern of the locale code fails. This means we have to constrain ourselves to not make step section names that could be plausible locale codes. (like what you did here with `locales.join('|')`

```js
router( `/start/:flowName/:stepName/:locale(${ locales.join('|')` );
```

In this case I still recommend we don't sequence multiple optional values since it opens up such a confusion in the code and in the runtime.